velox/exec/RowContainer.h

```diff
@@ -728,6 +726,48 @@ class RowContainer {
     }
   }
 
+  template <bool useRowNumbers, typename T>
+  static void extractValues(
+      const char* FOLLY_NONNULL const* FOLLY_NONNULL rows,
+      folly::Range<const vector_size_t*> rowNumbers,
+      int32_t numRows,
+      int32_t offset,
+      int32_t resultOffset,
+      FlatVector<T>* FOLLY_NONNULL result,
+      bool withNulls = false,
+      int32_t nullByte = 0,
+      uint8_t nullMask = 0) {
+    auto maxRows = numRows + resultOffset;
+    VELOX_DCHECK_LE(maxRows, result->size());
+
+    if (withNulls) {
```

Comment:
The idea before was to use templates to provide fast path for no-nulls code path. Did you find this not useful?

Reply:
The main reason for the consolidation is the duplication.

>  use templates to provide fast path for no-nulls code path

Per my understanding, the main difference between `extractValuesNoNulls` and `extractValuesWithNulls` is that no-nulls version has one less nullability check for each row:

https://github.com/facebookincubator/velox/blob/ce7f91f63f335ae63dbb18caca1f5b8bcea288d4/velox/exec/RowContainer.h#L874

https://github.com/facebookincubator/velox/blob/ce7f91f63f335ae63dbb18caca1f5b8bcea288d4/velox/exec/RowContainer.h#L840

If that's the problem here, how about templatizing `withNulls` like this:
```
template <bool withNulls, bool useRowNumbers, typename T>
static void extractValues(
    const char* FOLLY_NONNULL const* FOLLY_NONNULL rows,
    folly::Range<const vector_size_t*> rowNumbers,
    int32_t numRows,
    int32_t offset,
    int32_t resultOffset,
    FlatVector<T>* FOLLY_NONNULL result,
    int32_t nullByte = 0,
    uint8_t nullMask = 0) {
  auto maxRows = numRows + resultOffset;
  VELOX_DCHECK_LE(maxRows, result->size());

  if constexpr (withNulls) {
    result->mutableNulls(maxRows);
  }

  BufferPtr valuesBuffer = result->mutableValues(maxRows);
  auto values = valuesBuffer->asMutableRange<T>();

  auto assignNonNull = [&](const char* row, int32_t resultIndex) {
    result->setNull(resultIndex, false);
    if constexpr (std::is_same_v<T, StringView>) {
      extractString(valueAt<StringView>(row, offset), result, resultIndex);
    } else {
      values[resultIndex] = valueAt<T>(row, offset);
    }
  };

  for (int32_t i = 0; i < numRows; ++i) {
    const char* row;
    if constexpr (useRowNumbers) {
      auto rowNumber = rowNumbers[i];
      row = rowNumber >= 0 ? rows[rowNumber] : nullptr;
    } else {
      row = rows[i];
    }
    const auto resultIndex = resultOffset + i;
    if (row == nullptr) {
      result->setNull(resultIndex, true);
    } else if constexpr (withNulls) {
      if (isNullAt(row, nullByte, nullMask)) {
        result->setNull(resultIndex, true);
      } else {
        assignNonNull(row, resultIndex);
      }
    } else {
      assignNonNull(row, resultIndex);
    }
  }
}
```

This would make the generated code equivalent to the original version.

Reply:
Sounds good to me. Thanks.